Use CachedValue property, rather than IsUpdated to trigger UI updates

pboyer opened the pull request:
### Purpose

**`NodeModel.IsUpdated`, `NodeModel.cachedMirrorData`, and `NodeModel.CachedValue` are all used in slightly different ways and overlap in their meaning.** `IsUpdated` is meant primarily to indicate if a `NodeModel` was involved in a run, but it is also used to indicate if a `NodeModel` has had it's `cachedMirrorData` updated. These were the cause of many duplicate UI updates as `NodeView` used this property to determine if the `PreviewControl` should be updated. Furthermore, `CachedValue` is quite useful, but confusingly addressed throughout the code base. In this PR, I attempted to consolidate `CachedValue` and clarify some of the naming throughout the code base. 

Note, **I've removed the way that the `IsUpdated` setter mutates the `cachedMirrorData` field on `NodeModel`.** Consider the following code:

```
nodeModel.CachedValue = foo;
nodeModel.IsUpdated = true;
```

This code would result in `NodeModel.CachedValue` returning `null`. That is baffling behavior, and the cause of many recent defects. Perhaps there is some other purpose in mind here?

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@ikeough 

### FYI

@Benglin @ke-yu I've made a few subtle changes here. Maybe you can take a look to ensure that I've not done something dumb. ;)

pboyer commented on src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:
```diff
@@ -501,7 +501,7 @@ protected override void OnNodePropertyChanged(object sender, PropertyChangedEven
             switch (e.PropertyName)
             {
-                case "IsUpdated":
+                case "CachedValue":
```
This component is erroneously using `IsUpdated`, which could potentially cause it to initiate multiple rapid and potentially unnecessary updates.

pboyer commented on src/Libraries/CoreNodeModels/ColorRange.cs:
```diff
@@ -59,7 +59,7 @@ protected virtual void InitializePorts()
         {
-            if (e.PropertyName != "IsUpdated")
+            if (e.PropertyName != "CachedValue")
                 return;
```
This component is also erroneously using `IsUpdated`, which could potentially cause it to initiate multiple rapid and potentially unnecessary updates.

Benglin replied:
I guess this wasn't quite erroneous, but rather we did not have `CachedValue` to depend on before your change. Yes the frequency can be high, good fix there!

pboyer commented on src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/WatchImage.cs:
```diff
@@ -41,7 +41,7 @@ public void CustomizeView(WatchImageCore nodeModel, NodeView nodeView)
         private void NodeModelOnPropertyChanged(object sender, PropertyChangedEventArgs args)
         {
-            if (args.PropertyName != "IsUpdated") 
+            if (args.PropertyName != "CachedValue") 
```
This component is erroneously using `IsUpdated`, which could potentially cause it to initiate multiple rapid and potentially unnecessary updates.

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1612,19 +1611,19 @@ internal void RequestValueUpdateAsync(IScheduler scheduler, EngineController eng
-                var task = asyncTask as QueryMirrorDataAsyncTask;
-                cachedMirrorData = task.MirrorData;
+                this.CachedValue = task.MirrorData;
             }
```
This changes a few things:

1. We now actually clean up the event registration
2. We check the task is of the correct type (unlikely but possible)
3. We remove lock and use the `CachedValue` setter that correctly locks and emits the appropriate `CachedValue` PropertyChanged event.

Benglin replied:
I guess there's no need to un-register for the event, but good thing to do. For incorrect task type, instead of silently handle it, it'd be more appropriate to throw an exception wouldn't it?

Benglin commented:
The original intention of `IsUpdated`, like you mentioned, is more like `IsInvolvedInExecution`. I think you're keeping `IsUpdated` to give it the same meaning that *this node involved in the most recent execution*, and then use `RaisePropertyChanged("CachedValue")` to mark the time when the actual `MirrorData` is obtained. Right? If that is the case I think no UI component should be listening to `IsUpdated` property-changed event, because all they care is when the actual data has arrived. Perhaps `IsUpdated` is useful for Reach server to determine which are the nodes whose results are to be sent back.

Benglin commented:
Looks good otherwise, need to run this through CI though...

ikeough commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -454,16 +463,6 @@ public bool IsUpdated
             set
             {
                 isUpdated = value;
-                if (isUpdated)
```
Nice. There should be a law about property setters not setting other properties.

pboyer replied:
https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards#mutation

ikeough commented:
@pboyer This is great. Is there a way that we can enforce people NEVER using `IsUpdated`. It seems like you've removed many of the places that it is used "erroneously". What if `IsUpdated` just becomes internal? Is that possible?

pboyer commented:
Thanks @Benglin & @ikeough. I updated the PR, changing `IsUpdated` to `WasInvolvedInExecution` and made the field internal. I also cleaned up the docs a bit.
